src/main/java/com/r307/arbitrader/service/model/PaperTradeService.java

```diff
@@ -0,0 +1,116 @@
+
+public class PaperTradeService extends BaseExchangeService<PaperExchange> implements TradeService {
+
+    int orderId=0;
+
+    TradeService tradeService;
+
+    List<LimitOrder> openOrders=new ArrayList<>();
+
+    UserTrades userTrades = new UserTrades (new ArrayList<>(), Trades.TradeSortType.SortByTimestamp);
+
+    public PaperTradeService(PaperExchange exchange, TradeService tradeService) {
+        super(exchange);
+        this.tradeService=tradeService;
+        Timer timer = new Timer();
+        timer.schedule(new TimerTask() {
+            @Override
+            public void run() {
+                closeOrders();
+            }
+        }, 0, 1000);
+    }
+
+    public OpenOrders getOpenOrders() {
+        return new OpenOrders(openOrders);
+    }
+
+    public OpenOrders getOpenOrders(OpenOrdersParams params) {
+        return new OpenOrders(openOrders);
+    }
+
+    public String placeLimitOrder(LimitOrder limitOrder) {
+        orderId++;
+        LimitOrder limit = new LimitOrder(limitOrder.getType(),
+            limitOrder.getOriginalAmount(),
+            limitOrder.getInstrument(),
+            orderId+"",
+            new Date(),
+            limitOrder.getLimitPrice());
+        openOrders.add(limit);
+        return limit.getId();
+    }
+
+    public boolean cancelOrder(String orderId) {
+        return openOrders.removeIf(obj -> obj.getId().equals(orderId));
+    }
+
+    public UserTrades getTradeHistory(TradeHistoryParams params) {
+        return userTrades;
+    }
+
+    public TradeHistoryParams createTradeHistoryParams() {
+        return tradeService.createTradeHistoryParams();
+    }
+
+    public OpenOrdersParams createOpenOrdersParams() {
+        return tradeService.createOpenOrdersParams();
+    }
+
+    public void verifyOrder(LimitOrder limitOrder) {
+         //DO NOTHING
+    }
+
+    public void verifyOrder(MarketOrder marketOrder) {
+        //DO NOTHING
+    }
+
+    public Collection<Order> getOrder(String... orderIds) {
+        return openOrders.stream().filter(order -> Arrays.asList(orderIds).contains(order.getId())).collect(Collectors.toList());
+    }
+
+    public Collection<Order> getOrder(OrderQueryParams... orderQueryParams) {
+        List<String> orderIds = Arrays.stream(orderQueryParams).map(OrderQueryParams::getOrderId).collect(Collectors.toList());
+        return openOrders.stream().filter(order -> orderIds.contains(order.getId())).collect(Collectors.toList());
+    }
+
+
+    private void closeOrders() {
+        for(LimitOrder order: openOrders) {
+            userTrades.getUserTrades().add(new UserTrade(Order.OrderType.ASK, order.getOriginalAmount(),
+                order.getInstrument(),
+                order.getLimitPrice(),
+                order.getTimestamp(),
+                order.getId(),
+                order.getId(),
+                order.getFee(),
+                Currency.USD,
+                order.getUserReference()));
+        }
+        openOrders.clear();
```

Comment:
I got an exception trying to close a position. The orders when I open the position both get an ID of 1, then they close and the bot can see that they "filled". On close we try to look up the order volume by the order ID on the exchange - not all exchanges support this - but we catch a `NotAvailableFromExchangeException` in that case. What actually happened was it called `getOrder("1")` and I presume it got an empty `Collection` back.

The easiest thing I think you could try is making the `getOrder()` methods just throw `NotAvailableFromExchangeException` and that might fix it. If that happens we fall back to trying the account balance in the cryptocurrency, and if even that fails we fall back to a default value.

If you wanted to put in a little more work for a nicer solution you could keep both the open orders and the completed orders and make `getOrders()` capable of looking up closed orders by ID as well as open ones.

Reply:
Indeed, the PaperTradeService was only keeping track of the open oders and of the trade history. All orders are now kept in memory.